File: app/src/main/java/com/whiteside/insta/LoginActivity.java
package com.whiteside.insta;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;

public class LoginActivity extends AppCompatActivity {

    private TextInputEditText password, email;
    private Button login;

    private FirebaseAuth auth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        auth = FirebaseAuth.getInstance();

        email = findViewById(R.id.login_email);
        password = findViewById(R.id.password);
        login = findViewById(R.id.login);
    }

    public void loginClicked(View view) {
        final String password = this.password.getText().toString(),
                username = this.email.getText().toString();

        if (username.equals("")) {
            this.email.setError("E-mail is required");
        } else if (password.equals("")) {
            this.password.setError("Password is required");
        } else {
            login.setEnabled(false);
            auth.signInWithEmailAndPassword(username, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                @Override
                public void onComplete(@NonNull Task<AuthResult> task) {
                    if (task.isSuccessful()) {
                        successLogin();
                    } else {
                        failedLogin();
                    }
                    login.setEnabled(true);
                }
            });
        }
    }

    private void successLogin() {
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                Intent intent = new Intent(LoginActivity.this, FeedActivity.class);
                startActivity(intent);
                finish();
            }
        }, 500);
    }

    private void failedLogin() {
        ConnectivityManager cm =
                (ConnectivityManager) this.getSystemService(Context.CONNECTIVITY_SERVICE);

        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        boolean isConnected = activeNetwork != null &&
                activeNetwork.isConnectedOrConnecting();

        if (!isConnected) {
            internetError();
        } else {
            Toast.makeText(this, "username or password is not correct", Toast.LENGTH_LONG).show();
        }
    }

    private void internetError() {
        AlertDialog.Builder dialog = new AlertDialog.Builder(this);

        dialog.setTitle("Internet Problem");
        dialog.setMessage("There is an internet problem");

        dialog.setPositiveButton("Try again", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                loginClicked(login.getRootView());
                dialog.cancel();
            }
        });

        dialog.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        }).show();
    }

    public void registerClicked(View view) {
        Intent intent = new Intent(this, RegisterActivity.class);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/com/whiteside/insta/LoginActivity.kt
package com.whiteside.insta

import Model.Profile
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProviders
import com.facebook.AccessToken
import com.facebook.CallbackManager
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import com.facebook.login.widget.LoginButton
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.common.api.ApiException
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FacebookAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.firestore.*
import com.whiteside.insta.databinding.ActivityLoginBinding
import java.io.ByteArrayOutputStream
import java.util.*

class LoginActivity : AppCompatActivity() {
    private val callbackManager = CallbackManager.Factory.create()
    private var facebookLogin: LoginButton? = null
    private var auth: FirebaseAuth? = null
    private var googleSignInClient: GoogleSignInClient? = null
    private var viewModel: LoginViewModel? = null
    private var bind: ActivityLoginBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bind = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(bind!!.root)
        viewModel = ViewModelProviders.of(this).get(LoginViewModel::class.java)
        facebookLogin = bind!!.facebookLogin
        auth = FirebaseAuth.getInstance()
        registerFacebookCallback()
        setGoogleLoginListener()
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == GOOGLE_SIGN_IN) {
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            try {
                val account = task.getResult(ApiException::class.java)!!
                val credential = GoogleAuthProvider.getCredential(account.idToken, null)
                auth!!.signInWithCredential(credential)
                        .addOnFailureListener { e: Exception ->
                            e.printStackTrace()
                            com.whiteside.insta.GoogleSignIn.getGoogleSignInClient(this).signOut()
                        }
                        .addOnSuccessListener { authResult: AuthResult? -> checkUserInDatabase() }
            } catch (e: ApiException) {
                e.printStackTrace()
            }
        } else {
            callbackManager.onActivityResult(requestCode, resultCode, data)
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    private fun registerFacebookCallback() {
        facebookLogin!!.setPermissions(EMAIL, PUBLIC_PROFILE)
        facebookLogin!!.registerCallback(callbackManager, object : FacebookCallback<LoginResult> {
            override fun onSuccess(loginResult: LoginResult) {
                connectUserToFirebase(loginResult.accessToken)
            }

            override fun onCancel() {
                Toast.makeText(this@LoginActivity, "Login canceled", Toast.LENGTH_SHORT).show()
            }

            override fun onError(e: FacebookException) {
                e.printStackTrace()
                Toast.makeText(this@LoginActivity, "An error occurred while login", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun connectUserToFirebase(accessToken: AccessToken) {
        auth!!.signOut()
        val credential = FacebookAuthProvider.getCredential(accessToken.token)
        auth!!.signInWithCredential(credential)
                .addOnSuccessListener { authResult: AuthResult? -> checkUserInDatabase() }
                .addOnFailureListener { e: Exception ->
                    e.printStackTrace()
                    Toast.makeText(this@LoginActivity, "an error occurred", Toast.LENGTH_SHORT).show()
                    LoginManager.getInstance().logOut()
                }
    }

    private fun checkUserInDatabase() {
        viewModel!!.profileGetter.observe(this, { profile: Profile? ->
            if (profile != null) {
                successLogin()
            } else {
                addUserToDatabase()
            }
        })
        viewModel!!.getProfile(auth!!.uid)
    }

    private fun addUserToDatabase() {
        val profile = Profile()
        profile.name = Objects.requireNonNull(auth!!.currentUser)!!.displayName
        profile.bio = "Bio"
        profile.email = auth!!.currentUser!!.email
        profile.setFriends(ArrayList())
        profile.setFriendRequests(HashMap())
        val bitmap = BitmapFactory.decodeResource(resources, R.drawable.profile)
        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 50, stream)
        profile.profilePic = Blob.fromBytes(stream.toByteArray())
        viewModel!!.profileSetter.observe(this, { profile1: Profile? -> successLogin() })
        viewModel!!.setProfile(profile, auth!!.uid)
    }

    private fun successLogin() {
        val intent = Intent(this@LoginActivity, WallActivity::class.java)
        startActivity(intent)
        finish()
    }

    fun setGoogleLoginListener() {
        googleSignInClient = com.whiteside.insta.GoogleSignIn.getGoogleSignInClient(this)
        bind!!.signInButton.setOnClickListener { v: View? ->
            val signInIntent = googleSignInClient!!.signInIntent
            startActivityForResult(signInIntent, GOOGLE_SIGN_IN)
        }
    }

    companion object {
        private const val EMAIL = "email"
        private const val PUBLIC_PROFILE = "public_profile"

        //To add feature for suggesting to add his photos
        private const val USER_PHOTOS = "user_photos"
        private const val GOOGLE_SIGN_IN = 1003
    }
}